### ISSUE TITLE
feat: `ibis.select()`

### ISSUE THREAD

**Issue by NickCrews** [ibis-project/ibis#9324]
### Is your feature request related to a problem?

In mismo, [I wrap many of sklearns clustering metrics](https://github.com/NickCrews/mismo/blob/94087eb419af31688c5f73ed967f50b8b71a47c4/mismo/cluster/_eval.py#L154-L177). The relevant code is simplified here:

```python
def rand_score(labels_true: ir.Table, labels_pred: ir.Table) -> float:
    """Compute the Rand Index between two clusterings.

    The two input tables must have columns "record_id" and "label",
    a map from record ID to cluster label. They must have the same record IDs.

    See [sklearn.metrics.rand_score][] for more information.
    """
    from sklearn import metrics as _metrics
    labels_true, labels_pred = _to_numpy_labels(labels_true, labels_pred)
    return _metrics.rand_score(labels_true, labels_pred)


def _to_numpy_labels(
    labels_true: ir.Table, labels_pred: ir.Table
) -> tuple[np.ndarray, np.ndarray]:
    labels_true = labels_true["record_id", "label"]
    labels_pred = labels_pred["record_id", "label"]
    labels_true_df = labels_true.order_by("record_id").to_pandas()
    labels_pred_df = labels_pred.order_by("record_id").to_pandas()
    if not labels_true_df.record_id.equals(labels_pred_df.record_id):
        raise ValueError("labels_true and labels_pred must be aligned")
    return labels_true_df.label.values, labels_pred_df.label.values
```

In sklearn's API, these functions all take two arrays of labels. Since they are numpy arrays, they can be implicitly aligned. However, since ibis columns can't be aligned, this presents a problem. Several possible APIs I've considered:

## A) `def f(a: ir.Table, b: ir.Table)`

This is what I have currently. each table must have columns "record_id" and "label". The tables can be aligned on the "record_id" column. But this isn't very ergonomic. Often, when users want to use these functions, both of the columns are in the same Table. So, they have to do something like

```python
labels_true = labeled.select("record_id", label=_.label_true)
labels_pred = labeled.select("record_id", label=_.component)
mismo.cluster.adjusted_rand_score(labels_true, labels_pred)
```

which is a little more awkward that I'd like.

##  B) `def f(t: ir.Table, a: str | ibis.Deferred, b: str | ibis.Deferred)`

This API makes more sense for the above use case. People could do `mismo.cluster.adjusted_rand_score(labeled, "label_true", "component")`. But it still is sorta ugly. I am passing in the entire table, which isn't great, I want the API to only take the data that is required. This is currently possible with no changes in ibis.

## C) `def f(a: ir.Column , b: ir.Column)`

This would be my ideal API. Then the use case above could be simplified to `mismo.cluster.adjusted_rand_score(labeled.label_true, labeled.component)`. We are assuming that users of this API only are passing in two columns that can be correlated.

However, this is impossible for me to implement without reaching into the internals of ibis. I could do something like `ibis.expr.analyze.find_parent_relation(a)` and then doing some kung fu to merge a and b into the same table.

### What is the motivation behind your request?

see above

### Describe the solution you'd like

a top-level api of `ibis.select(*args, **kwargs)` with the same semantics as `Table.select`. This would allow me to combine these two Expressions into a single Table without needing a handle on the parent relation. This is similar to [polars.select()](https://docs.pola.rs/py-polars/html/reference/expressions/api/polars.select.html#polars.select).

It would raise an error if the arguments can't be resolved to a single base Relation. This could use the logic that already exists inside Table.select.

Alternatively, the API could be added into `ibis.table` somehow?

Then, I could use it as

```python
def rand_score(labels_true: ir.Column, labels_pred: ir.Column) -> float:
    """Compute the Rand Index between two clusterings.

    See [sklearn.metrics.rand_score][] for more information.
    """
    from sklearn import metrics as _metrics
    labels_true, labels_pred = _to_numpy_labels(labels_true, labels_pred)
    return _metrics.rand_score(labels_true, labels_pred)

def _to_numpy_labels(
    labels_true: ir.Column, labels_pred: ir.Column
) -> tuple[np.ndarray, np.ndarray]:
	t = ibis.select(labels_true.name("true"), labels_pred.name("pred"))
    df = t.to_pandas()
    return df.true.values, df.pred.values
```

Note how now the docstring can be much simpler and the function is much easier to understand.

### What version of ibis are you running?

main

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by jcrist** [comment]
This is pretty tricky to do generically, but is something that [has come up as well in `ibis-ml`](https://github.com/ibis-project/ibis-ml/blob/26b704a71f83ad0729f28e6eae05c7882f171701/ibis_ml/core.py#L56) for supporting passing in `X` & `y` in a number of different formats (and combining them into a meaningful table). That use case is a bit simpler since `X` is always a table (and with a table reference the code to make it work is simpler). I can sympathize with what you're asking for though.

A few random thoughts:

1. I don't love the name `ibis.select`, but also don't hate it. A few alternatives:
  - `ibis.concat` (similar to `pd.concat` or `pl.concat`, which is (AFAICT) a closer fit to what you're asking for than `pl.select`)
  - `ibis.hconcat` (like above, but only supports horizontal concatenation (since `ibis.union` already works for vertical))
  - `ibis.align`?
  - `ibis.combine`?
  - ?

2. I think we also want to support up to 1 table in the input. Perhaps a signature like:

```python
def this_new_api(*args: ibis.Column | ibis.Table) -> ibis.Table:
    """Horizontally combine aligned columns or tables into a single table"""
    # error nicely if there's more than one table in args
    # error nicely if all args cannot be merged correctly into a single valid table
```

3. This seems like an API that most users wouldn't need. I'd worry that a too obvious name may distract novice users into using this instead of constructing relations through more standard methods. Maybe not worth worrying about - we can't really stop users from writing bad/weird code, but we can make the correct paths more obvious/ergonomic.

**Comment by NickCrews** [comment]
> 1. I don't love the name `ibis.select`, but also don't hate it. A few alternatives:

my opinions on these names
- ibis.select: yeah it feels very SQLy, but the rest of ibis is more dataframey. I feel the same.
- ibis.concat: I strictly prefer hconcat over this. Too ambiguous.
- ibis.hconcat: my favorite I think. This makes me want to also support vconcat as an alias for union. Maybe even deprecate union?
- ibis.align: I do not like this, makes me think it is some sort of zipping, doesn't make me think of combining.
- ibis.combine: concat just seems strictly better than this.

> 2. I think we also want to support up to 1 table in the input.

If you already have a table, why can you not do `table.mutate(columns)`, optionally followed by a `.relocate()`?

> 3. This seems like an API that most users wouldn't need.

Hmm, I don't think I quite understand the concern here. Why would this method be inferior? I agree most of the time they could probably use `ibis.table()/ibis.memtable()`, but why *should* they use those APIs instead?

I think now that I just found #9309, I think that is my favorite path forward by far.

**Comment by cpcloud** [comment]
How does `ibis.table` fit into this issue? We should definitely avoid stuffing **more** functionality into `ibis.table`. We're already considering giving it two distinct behaviors. Adding another one is not a good idea.

**Comment by NickCrews** [comment]
this feature request is about constructing a Table from some columns and scalars. In #9309, we are considering making `ibis.table()` create a Table from python objects. I am suggesting extending that to also accept ibis scalars and columns.

> two distinct behaviors

just to be sure, you mean 1. a table unbound to data and 2. a table bound to data?

**Comment by cpcloud** [comment]
> just to be sure, you mean 1. a table unbound to data and 2. a table bound to data?

That's correct.

**Comment by NickCrews** [comment]
I think my proposal fits into those two modes? In both those modes we have to take the passed input and infer a schema from it. The passed ibis objects themselves may be bound to data, or unbound, would affect the implementation of table()? Am I missing something that makes these two use cases different?

**Comment by cpcloud** [comment]
> Am I missing something that makes these two use cases different?

I think so.

> this feature request is about constructing a Table from some columns and scalars.

Yes, but that describes most of what Ibis does.

#9309 is about constructing a **leaf** table node from atomic sources, one source being an in memory table, the other source being a table name and a schema.

A leaf table node cannot be constructed from an arbitrary set of other column expressions, because those expressions themselves have other table nodes as leaves.

It really doesn't make sense to stuff more functionality other than what's being proposed, into #9309.

**Comment by thomasgilgenast** [comment]
I stumbled into this and #9309 with a similar but slightly different use case. I was rewriting some polars code to use ibis, and was missing something like polars `with_columns()` to add a new column to an existing ibis Table using data from an in-memory numpy array. For this `ibis.hconcat()` would be a reasonable fit for me I think. The solutions proposed here and in #9309 sound more comprehensive than `ibis.hconcat()`, but in case it's helpful for others who stumble into this thread looking for `ibis.hconcat()`, here is a quick-and-dirty workaround I am using in the meantime:

```python
import uuid
import ibis
import ibis.expr.types.relations as ir

def hconcat(t1: ir.Table, t2: ir.Table) -> ir.Table:
    assert not set(t1.columns) & set(t2.columns), "tables have overlapping columns"
    join_key = "_hconcat_idx_" + uuid.uuid4().hex[:7]
    assert join_key not in t1, f"t1 has a {join_key} column"
    assert join_key not in t2, f"t2 has a {join_key} column"
    return ibis.join(
        t1.mutate(**{join_key: ibis.row_number()}),
        t2.mutate(**{join_key: ibis.row_number()}),
        join_key,
    ).select([x for x in t1.columns] + [x for x in t2.columns])
```

I was able to combine this with `ibis.memtable()` to add a column to an existing table using data from a numpy array:

```python
new_table = hconcat(old_table, ibis.memtable(numpy_array, columns=["new_column_name"]))
```

**Comment by NickCrews** [comment]
> leaf table node

Thanks for that framework, I hadn't thought of partitioning Table expressions into being either leaf or non-leaf. Up to this point we were thinking that ibis.table() should only be responsible for creating leaf nodes. But is that only baggage from its history? From a users perspective, I would love it if ibis.table() just gave me a Table back, I don't want to have to worry about if it is leaf or non-leaf. Am I missing some footgun here why that's a bad idea? 

> It really doesn't make sense to stuff more functionality

Just to be clear, are you concerned with

1. Putting too much in one PR. Sorry, I wasn't very clear, I don't think we need to stuff this into that one PR, just I want to leave the door open to extend the ibis.table() API in a future PR. I think it actually is already at that point, just wanted to put it on your radar.
2. The semantics are wrong, we shouldn't ever allow passing ibis objects into ibis.table()

**Comment by cpcloud** [comment]
We should avoid god functions that accept huge varieties of inputs and supposedly "just do the right thing" because that's basically never the case, or it comes with some awful maintenance trade-off.

`ibis.table` is only for creating leaves and that's intentional. It's not baggage, and it's not related to PR size.

Keeping APIs limited to doing a very small number of things is how we keep ibis composable and maintainable long term.

**Comment by NickCrews** [comment]
Thanks Philip, I can agree with all that. I can shut up now ;)

OK, it seems like we are settling on a `hconcat(<ibis values>)` API. @thomasgilgenast brings up an interesting question: Do we want to only support expressions from 1 base relations, or do we want to support multiple base relations, joined by implicit row number? If we allow it, I think it is VERY likely that (especially beginner) users will expect Pandas/numpy-like semantics on row order being preserved, and getting footgunned when that isn't guaranteed. I think a better solution would be to add a `how="position"` option to `ibis.join()`

**Comment by cpcloud** [comment]
> Thanks Philip, I can agree with all that. I can shut up now ;)

I very much appreciate your continued engagement with the project :)

**Comment by NickCrews** [comment]
Actually, thinking about this more I think we should
1. Add `how="positional"` to `ibis.join`. I think we can/should keep that as the one API for combining two different relations together
2. keep this as `ibis.select()`, and it should be totally symmetrical to Table.select: all arguments need to be scalars and columns from the same relation. You can't pass a Table into it: if you have access to a Table, you should use the Table.select() API instead. I want to keep the name as ibis.select() to make it very clear this is analogous to Table.select().
